src/main/java/uk/ac/ebi/ampt2d/accession/common/accessioning/BasicAccessioningService.java

```diff
@@ -0,0 +1,136 @@
+import uk.ac.ebi.ampt2d.accession.common.utils.DigestFunction;
+import uk.ac.ebi.ampt2d.accession.common.utils.HashingFunction;
+import uk.ac.ebi.ampt2d.accession.common.generators.AccessionGenerator;
+import uk.ac.ebi.ampt2d.accession.common.persistence.DatabaseService;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * A service that provides accessions for objects
+ *
+ * @param <MODEL>
+ * @param <ACCESSION>
+ * @param <HASH>
+ */
+public class BasicAccessioningService<MODEL, HASH, ACCESSION> implements InitializingBean,
+        AccessioningService<MODEL, ACCESSION> {
+
+    private AccessionGenerator<MODEL, ACCESSION> accessionGenerator;
+
+    private DatabaseService<MODEL, HASH, ACCESSION> dbService;
+
+    private final DigestFunction<MODEL> digestFunction;
+
+    private HashingFunction<HASH> hashingFunction;
+
+    /**
+     * Basic accessioning service constructor, requires an accession generator, a database service and a hashing
+     * function, it also allows the registering of a init function.
+     *
+     * @param accessionGenerator
+     * @param dbService
+     * @param hashingFunction
+     */
+    public BasicAccessioningService(AccessionGenerator<MODEL, ACCESSION> accessionGenerator,
+                                    DatabaseService<MODEL, HASH, ACCESSION> dbService,
+                                    DigestFunction<MODEL> digestFunction,
+                                    HashingFunction<HASH> hashingFunction) {
+        this.accessionGenerator = accessionGenerator;
+        this.dbService = dbService;
+        this.digestFunction = digestFunction;
+        this.hashingFunction = hashingFunction;
+    }
+
+    /**
+     * Get accessions for a list of messages. It looks for the object's accessions in a repository, and it they don't
+     * exist, generate new ones, storing them in the repository
+     *
+     * @param messages
+     * @return
+     */
+    @Override
+    public Map<ACCESSION, MODEL> getAccessions(List<? extends MODEL> messages) {
+        Map<HASH, MODEL> hashToMessages = mapHashOfMessages(messages);
+        Map<HASH, ACCESSION> existingAccessions = dbService.getExistingAccessions(hashToMessages.keySet());
+        Map<HASH, MODEL> newMessages = filterNotExists(hashToMessages, existingAccessions);
+
+        Map<ACCESSION, MODEL> accessions = joinExistingAccessionsWithMessages(existingAccessions, hashToMessages);
+        if (!newMessages.isEmpty()) {
+            accessions.putAll(generateAccessions(newMessages));
+        }
+        return accessions;
+    }
+
+    /**
+     * Digest messages, hash them and map them. If Two messages have the same hash keep the first one.
+     *
+     * @param messages
+     * @return
+     */
+    private Map<HASH, MODEL> mapHashOfMessages(List<? extends MODEL> messages) {
+        return messages.stream().collect(Collectors.toMap(digestFunction.andThen(hashingFunction), e -> e, (r, o) -> r));
+    }
```

Comment:
digestFunction.andThen(hashingFunction) does this make hash and accession the same ?

Reply:
Nope, digest generates the message, hashing calculates the hash from the message.

Reply:
SingleAccessionGenerator also does the same for generating accessions.

Reply:
Not exactly, SingleAccessionGenerator needs a function to return an ID, and there is a constructor that lets you create a SingleAccessionGenerator that gets the ID making the hash of the digested object.